added test cases. Added containsAll function. Resolved issues in checkTupleArray method in TypeCheckVisitor

RamIowaStateUniversity opened the pull request:
Final commit for program-analysis feature. Added test cases. Added containsAll method but removed addAll and removeAll as I do not think there is any performance implication when compared to union and difference.

Performance implication occurs with the existing foreach statement to iterate sets, as we convert set to arrays and then iterate it.

hridesh commented on src/java/boa/compiler/SymbolTable.java:
```diff
@@ -192,6 +192,7 @@
 
 		// set functions
 		globalFunctions.addFunction("contains", new BoaFunction(new BoaBool(), new BoaType[] { new BoaSet(new BoaTypeVar("V")), new BoaTypeVar("V") }, "${0}.contains(${1})"));
+		globalFunctions.addFunction("containsAll", new BoaFunction(new BoaBool(), new BoaType[] { new BoaSet(new BoaTypeVar("V")), new BoaSet(new BoaTypeVar("V"))}, "${0}.containsAll(${1})"));
```
Should the function naming conventions be similar to current naming conventions used within Boa?

psybers replied:
This is a good point.  But perhaps a better point is: what **is** our naming convention?  The majority seem to follow no uppercase, no underscore.  But I see one with an underscore.  So should this be 'containsall'?  Also, similar issue with the other new function 'getValue'.

psybers commented:
Hridesh's earlier comments still stand: these test cases are helpful, but what we need are **simple** test cases that test a single new feature in isolation.  Then when the test suite runs, we know if a test fails what exactly the feature is that failed.  Look at the tests in the typecheck, parser, codegen dirs for examples.

RamIowaStateUniversity commented:
Addressed both naming convention and test case issue.

On Sun, Jan 29, 2017 at 2:03 PM, Robert Dyer <notifications@github.com>
wrote:

> Hridesh's earlier comments still stand: these test cases are helpful, but
> what we need are *simple* test cases that test a single new feature in
> isolation. Then when the test suite runs, we know if a test fails what
> exactly the feature is that failed. Look at the tests in the typecheck,
> parser, codegen dirs for examples.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/boalang/compiler/pull/93#issuecomment-275941713>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/APaRopyurKwNvjala2-ch6T_XZDc8HQDks5rXPCmgaJpZM4LwZHp>
> .
>

hridesh commented:
Some of these new tests might need to be verified. For example, the latest change uses "getvalue" as opposed to "getValue". However, some tests e.g. test/traversal/getValue-current-traversal.boa still uses the older keyword.

psybers commented:
Are the new tests not included with ‘ant test’? If they are and they are invalid, then CI would have failed.

hridesh commented:
I looked over the CI, and the new tests would need to be included. Currently, new tests are not running as far as I can see.

RamIowaStateUniversity commented:
Added TestTraversalGood and TestTraversalBad.
Added test cases to build.xml.

Thanks,
Ram

On Sun, Jan 29, 2017 at 9:58 PM, Hridesh Rajan <notifications@github.com>
wrote:

> I looked over the CI, and the new tests would need to be included.
> Currently, new tests are not running as far as I can see.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/boalang/compiler/pull/93#issuecomment-275975643>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/APaRomSNVBjxLMf6VbLNknMnmVHUeiLaks5rXV_YgaJpZM4LwZHp>
> .
>

psybers commented on src/java/boa/compiler/visitors/TypeCheckingVisitor.java:
```diff
@@ -1525,13 +1525,15 @@ protected boolean checkTupleArray(final List<BoaType> types) {
 		if (types == null)
 			return false;
 
-		final String type = types.get(0).toBoxedJavaType();
```
What is this change?  It looks like you are undoing some code cleanup I did months ago?

RamIowaStateUniversity commented:
I have reverted the change. But there are some issues with tuple, like if a
tuple contains only fields of same type, then an error is thrown.

For example, the below case throws an error.

type T_inout= {in: set of string, out: set of string};

I have changed the example to contain one more different data type, but we
need to create an issue and look into what is happening. With my version of
typechecking visitor, the above case does not thrown an error.

Thanks,
Ram

On Mon, Jan 30, 2017 at 8:40 AM, Robert Dyer <notifications@github.com>
wrote:

> *@psybers* requested changes on this pull request.
> ------------------------------
>
> In src/java/boa/compiler/visitors/TypeCheckingVisitor.java
> <https://github.com/boalang/compiler/pull/93#pullrequestreview-19094294>:
>
> > @@ -1525,13 +1525,15 @@ protected boolean checkTupleArray(final List<BoaType> types) {
>  		if (types == null)
>  			return false;
>
> -		final String type = types.get(0).toBoxedJavaType();
>
> What is this change? It looks like you are undoing some code cleanup I did
> months ago?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/boalang/compiler/pull/93#pullrequestreview-19094294>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APaRom60vxI7JCbrHU57lH8Oh2iz_sdxks5rXfZkgaJpZM4LwZHp>
> .
>

psybers commented:
Well, then that is a bug in that logic.  The original code had a logic bug in it because the comparison of 2 strings was done with '==' instead of .equals().  So when I cleaned it up, I fixed that logic error.  What this method says is, if every field in the tuple is the same type it returns false (that was the intended logic in both versions).  So it is behaving somewhat rational in that sense - your example is 2 fields of the same type, so it doesnt mark it as a tuple (instead marking it as an array).  There should be conversions from array to tuple/vice versa where it makes sense.  So I think you need to file that as a new issue.

hridesh commented:
I am merging this with the expectation that we will handle enhancement of documentation to describe these features as part of another pull request.
